alternator/error.hh

```diff
@@ -27,10 +30,13 @@ public:
     status_type _http_code;
     std::string _type;
     std::string _msg;
-    api_error(std::string type, std::string msg, status_type http_code = status_type::bad_request)
+    std::optional<lw_shared_ptr<rjson::value>> _json; // shared pointer needed as api_error copy constructor is required
+    api_error(std::string type, std::string msg, status_type http_code = status_type::bad_request,
+    std::optional<lw_shared_ptr<rjson::value>> json = std::nullopt)
         : _http_code(std::move(http_code))
         , _type(std::move(type))
         , _msg(std::move(msg))
+        , _json(std::move(json))
```

Comment:
sorry, I accidentally stashed this change in server.cc, will add it.

Reply:
Ok, maybe after I'll see it I'll understand better. Did DynamoDB added to the protocol a way to return an item as part of an error, not just an error? Did you see it documented somewhere?

Reply:
Yes, since it's still (condition) error this feature changes error class and dynamod does return non-200 code and adds  values json there. 

AFAICS it's not really documented I had to wireshark it to see that's inside "Item" object.

This can be moved to separate commit.

Reply:
Yes, please move it to a separate commit explaining the new layout of error messages. The original format documented in https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html had "__type" and "message" keys. What is the new key (I guess I'll see it when I see your change to server.cc)? Is there no trace of it in any DynamoDB documentation? That's so weird.

Reply:
I still can't find any better documentation, but I found this discussion: https://github.com/boto/boto3/issues/2657 on TransactionCanceledException, which suggests that in addition to the "__type" and "Message", the error message can have additional keys and values  - in that case they have the key "CancellationReasons" with an object as a value.

So perhaps we need in api_error the general ability to add any further keys and values, and in ConditionCheckException specifically add a parameter "item" (not "json") which sets the "Item" key in the error message. When we later implement transactions, we'll need to support TransactionCanceledException and this will set the "CancellationReasons" key - not the "Item" key.

So perhaps api::error() should just have a JSON map variable, and if it's not empty, the individual entries from this map will be inserted into the error message JSON?